text/0000-pub-sub.md

```diff
@@ -0,0 +1,170 @@
+
+- Cross tenant events exchange is assumed in the future, but not in the scope of MVP (v1).
+
+- Integration with external/third party systems should be allowed (not in scope of MVP, but the basis for that must be provided)
+
+### Target Solution Architecture
+
+#### Solution Overview
+- A new module is introduced, `mod-pubsub`. It is responsible for maintaining the registrations of event types and for coordinating distribution of events to the appropriate subscribers. Folio modules may publish events and/or subscribe to events of specific event types. It is the responsibility of publisher modules to define the event types that they will provide, which is accomplished through the use of an event descriptor. Note that it is possible for multiple publishers to register for the same event type. However, the event type may only be defined once. Subscribing modules will register themselves to specific event types by providing a URI that will be invoked by the event manager when pushing events of matching event types.  Event subscription is done on a per tenant basis.
+
+- The `mod-pubsub` module is also responsible for maintaining a trail of activities: event publication; subscription distribution; errors; non-delivery; etc. By using the correlationID embedded in events it is possible to reconstruct the sequence of operations that form a particular workflow.
+
+- By configuring specific event subscriptions for each tenant it is possible to implement specific cases of inter-app coordination.
+
+- The `mod-pubsub` module is implemented using a message queue (Kafka) which will provide the necessary persistence.
+
+#### High-level solution structure
+There is a global FOLIO interface `pubsub` defined in the scope of this solution which provides APIs to ensure the functioning of the subsystem:
+- API for event type registration by publishers and subscribers. This API must be called by back-end modules.
+  - `POST /pubsub/eventtypes` - to register an event type 
+  - `DELETE /pubsub/eventtypes` - to remove event type registration
+  - `POST /pubsub/eventtypes/declare/publisher` - to declare a publisher for a set of event types
+  - `DELETE /pubsub/eventtypes/declare/publisher` - to remove publisher declaration
+  - `POST /pubsub/eventtypes/declare/subscriber` - to declare a subscriber for a set of event types
+  - `DELETE /pubsub/eventtypes/declare/subscriber` - to remove subscriber declaration
+- API to retrieve existing registered event types.
+  - `GET /pubsub/eventtypes` - returns all registered event types
+- API to retrieve the registered subscriptions for a specified tenant and event type
+  - `GET /pubsub/eventtypes/{EVENT_TYPE}/subscribers` - returns all subscribers of the event type provided in the path variable
```

Comment:
Is `EVENT_TYPE` a UUID?  Name?

Reply:
Yes, you are right. This must be humanly readable like INVENTORY_INSTANCE_CREATED for example.

Reply:
Is the all capitals with underscores (with version postfix) the recommended convention for event type names, or is this only an example to aid this conversation?

Reply:
The proposal is not constraining in how event type are named. Even the use of version number suffixes is just a suggestion.

Reply:
Thanks for clarifying that is out of scope for this RFC.

Do you think that FOLIO is likely to provide guidance on this separately?

Reply:
> Do you think that FOLIO is likely to provide guidance on this separately?

It would be a good idea to establish such guidance for specific uses of the Folio platform: e.g. Folio LSP.